# Task

This unknown company, the book service is trying to compete against other services in the market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

## Goals of the study
- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating: look only at books with at least 50 ratings.
- Find the average number of text reviews among users who rated more than 50 books.

### Create a database connections

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
#creating queries for all the tables
query1 = ''' SELECT *
            FROM books
            LIMIT 5;
        '''
query2 = ''' SELECT *
            FROM authors
            LIMIT 5;
        '''
query3 = ''' SELECT *
            FROM publishers
            LIMIT 5;
        '''
query4 = ''' SELECT *
            FROM ratings
            LIMIT 5;
        '''
query5 = ''' SELECT *
            FROM reviews
            LIMIT 5;
        '''

In [3]:
books = pd.io.sql.read_sql(query1, con = engine)
authors = pd.io.sql.read_sql(query2, con = engine)
publishers = pd.io.sql.read_sql(query3, con = engine)
ratings = pd.io.sql.read_sql(query4, con = engine)
reviews = pd.io.sql.read_sql(query5, con = engine)

In [4]:
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Make an SQL query for each task, output the results, and describe your conclusions


### Find the number of books released after January 1, 2000

In [9]:
TASK1 = '''
SELECT
    COUNT(*)
FROM
    books
WHERE
    publication_date > '2000-01-01'; 
'''
TASK1 = pd.io.sql.read_sql(TASK1, con = engine)
TASK1

count
0    819

#### Conclusion

From this query, we can find that the number of books released after January 1st of the year 2000 is 780.

### Find the number of user reviews and the average rating for each book.

In [10]:
TASK2 = '''
SELECT
    books.title AS title,
    COUNT(reviews.book_id) AS count,
    AVG(ratings.rating) AS rating    
FROM 
    books
    INNER JOIN ratings ON books.book_id = ratings.book_id
    INNER JOIN reviews ON books.book_id = reviews.book_id
GROUP BY
    title
ORDER BY
    COUNT(reviews.book_id) DESC;
'''
TASK2 = pd.io.sql.read_sql(TASK2, con = engine)
TASK2

title  count    rating
0                              Twilight (Twilight  #1)   1120  3.662500
1                  The Hobbit  or There and Back Again    528  4.125000
2                               The Catcher in the Rye    516  3.825581
3    Harry Potter and the Prisoner of Azkaban (Harr...    492  4.414634
4    Harry Potter and the Chamber of Secrets (Harry...    480  4.287500
..                                                 ...    ...       ...
988  The Adventures of Tom Sawyer and Adventures of...      2  5.000000
989                                  History of Beauty      2  2.500000
990                                         Lysistrata      2  4.000000
991                   Executive Orders (Jack Ryan  #8)      2  3.500000
992  Dragons of Winter Night (Dragonlance: Chronicl...      2  4.000000

[993 rows x 3 columns]

#### Conclusion
From this query, we can see each book title and the count of reviews and average rating for each. For example, we can see that Twilight has the most amount of reviews having 1120 user reviews, but has an average rating of 3.6.

### Identify the publisher that has released the greatest number of books with more than 50 pages


In [11]:
TASK3 = '''
SELECT
    publishers.publisher AS publisher,
    COUNT(publishers.publisher) AS count
FROM 
    books
    INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE
    num_pages > 50
GROUP BY
    publisher
ORDER BY
    COUNT(publisher) DESC
LIMIT 5;
'''
TASK3 = pd.io.sql.read_sql(TASK3, con = engine)
TASK3

publisher  count
0             Penguin Books     42
1                   Vintage     31
2  Grand Central Publishing     25
3          Penguin Classics     24
4                    Bantam     19

#### Conclusion
From this query, we can see the publishers that have released the greatest number of books that have more than 50 pages. Right now, we can see that the publisher Penguin Books has released 42 books.

### Identify the author with the highest average book rating (books with at least 50 ratings)

In [12]:
TASK4 = '''
SELECT
    authors.author AS author,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    INNER JOIN ratings ON books.book_id = ratings.book_id
    INNER JOIN authors ON books.author_id = authors.author_id
GROUP BY
    author
HAVING
    COUNT(ratings.rating) > 50
ORDER BY
    AVG(rating) DESC
LIMIT 5;
'''
TASK4 = pd.io.sql.read_sql(TASK4, con = engine)
TASK4

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.288462
1                    Agatha Christie    4.283019
2  Markus Zusak/Cao Xuân Việt Khương    4.264151
3                     J.R.R. Tolkien    4.240964
4           Roald Dahl/Quentin Blake    4.209677

#### Conclusion

Within the top 5 authors who have the highest average book rating with at least 50 ratings, J.K. Rowling has an outstanding 4.2 rating as the top author.

### Find the average number of text reviews among users who rated more than 50 books

In [13]:
TASK5='''    
    select 
    avg(sub1.cnt)
    from
        ( 
        SELECT
            reviews.username AS username,
            COUNT(reviews.review_id) AS cnt
        FROM  
            reviews
        GROUP BY
            reviews.username     
            )sub1
    inner join (SELECT
       username AS username,
       COUNT(rating_id) AS cnt
    FROM  
        ratings
    GROUP BY
        username
    HAVING
       COUNT(rating_id) > 50) sub2 on (sub1.username = sub2.username)
'''
TASK5 = pd.io.sql.read_sql(TASK5, con = engine)
TASK5

avg
0  24.333333

#### Conclusion

With this final query, I was able to output the average number of text reviews among users who have rated more than 50 books. As we can see with the output, the average number of text reviews is 24.3.